src/tiled/scriptxmlfile.cpp

```diff
@@ -0,0 +1,306 @@
+        return {};
+    }
+
+    if (!refChild) {
+        ScriptManager::instance().throwError(QStringLiteral("Second argument is undefined"));
+        return {};
+    }
+
+    return checkIfNull(mNode.insertBefore(importNode(newChild), refChild->mNode));
+}
+
+ScriptXmlNode *ScriptXmlNode::insertAfter(ScriptXmlNode *newChild, ScriptXmlNode *refChild)
+{
+    if (!newChild) {
+        ScriptManager::instance().throwError(QStringLiteral("First argument is undefined"));
+        return {};
+    }
+
+    if (!refChild) {
+        ScriptManager::instance().throwError(QStringLiteral("Second argument is undefined"));
+        return {};
+    }
+
+    return checkIfNull(mNode.insertAfter(importNode(newChild), refChild->mNode));
+}
+
+ScriptXmlNode *ScriptXmlNode::replaceChild(ScriptXmlNode *newChild, ScriptXmlNode *oldChild)
+{
+    if (!newChild) {
+        ScriptManager::instance().throwError(QStringLiteral("First argument is undefined"));
+        return {};
+    }
+
+    if (!oldChild) {
+        ScriptManager::instance().throwError(QStringLiteral("Second argument is undefined"));
+        return {};
+    }
+
+    return checkIfNull(mNode.replaceChild(importNode(newChild), oldChild->mNode));
+}
+
+ScriptXmlNode *ScriptXmlNode::removeChild(ScriptXmlNode *oldChild)
+{
+    if (!oldChild) {
+        ScriptManager::instance().throwError(QStringLiteral("First argument is undefined"));
+        return {};
+    }
+
+    return checkIfNull(mNode.removeChild(oldChild->mNode));
+}
+
+ScriptXmlFile::ScriptXmlFile(const QDomDocument &document, QObject *parent)
+    : ScriptXmlNode(document, parent)
+{}
+
+ScriptXmlFile::ScriptXmlFile(const QString &source, QObject *parent)
+    : ScriptXmlFile(QDomDocument(), parent)
+{
+    QString error;
+    int line;
+    int column;
+    if (!document().setContent(source, false, &error, &line, &column)) {
+        auto message = QStringLiteral("Unable to parse xml: %1 (line %2, column %3)")
+                .arg(error).arg(line).arg(column);
+        ScriptManager::instance().throwError(message);
+
+        // Return mDocument to a valid state.
+        document().setContent({});
+    }
+}
+
+ScriptXmlNode *ScriptXmlFile::createNode(NodeType type, const QString &data)
+{
+    switch (type) {
+    case NodeType::Element:
+        return new ScriptXmlNode(document().createElement(data));
+    case NodeType::Text:
+        return new ScriptXmlNode(document().createTextNode(data));
+    case NodeType::CDATA:
+        return new ScriptXmlNode(document().createCDATASection(data));
+    case NodeType::Root:
+        ScriptManager::instance().throwError(QStringLiteral("Cannot use createNode to make XmlFile instances. Use the constructor instead."));
+        return nullptr;
+    default:
+        ScriptManager::instance().throwError(QStringLiteral("Cannot create node type"));
+        return nullptr;
+    }
+}
+
+QString ScriptXmlFile::writeToString(int indent) const
```

Comment:
Did just calling this `toString` conflict with the JavaScript to string conversion?